File: dec02/game2.mjs
// https://adventofcode.com/2018/day/2
// part 2

import { readFileSync } from 'fs';

const input = readFileSync('./input.txt', 'utf8');

function parse_color(rgb) {
    const tries = rgb.split(',').map(s => s.trim());
    const rgbval = tries.map(cval => cval.split(' ').reverse());
    return Object.assign(
        {red: 0, green: 0, blue: 0}, 
        Object.fromEntries(rgbval.map(v => [v[0], parseInt(v[1])]))
    );
}

function parse_colors(input) {
    return input.split(';').map(parse_color);
}

const games = input
    .split('\n').filter(line => line.trim().length > 0)
    .map(line => ({
        id: parseInt(line.match(/Game (\d+): /)[1]),
        colors: parse_colors(line.match(/: (.*)/)[1])
    }));


const maxcolors = {red: 12, green: 13, blue: 14};

const color_le = (c1, c2) => {
    for (const c in c1) {
        if (c1[c] > c2[c]) return false;
    }
    return true;
}

const mincubes = game => ({
    red: Math.max(...game.colors.map(c => c.red)),
    green: Math.max(...game.colors.map(c => c.green)),
    blue: Math.max(...game.colors.map(c => c.blue))
});

const abscubes = c => c.red * c.green * c.blue;

// 78669
console.log(
    games.map(mincubes).map(abscubes).reduce((acc, val) => acc + val, 0)
);

File: dec02/game2.py
# https://adventofcode.com/2018/day/1
# part 2
from __future__ import annotations
from dataclasses import dataclass, field


@dataclass
class Colors:
    red: int
    green: int
    blue: int

    def __le__(self, other):
        return self.red <= other.red and self.green <= other.green and self.blue <= other.blue

    def __abs__(self):
        return self.red * self.green * self.blue

@dataclass
class Game:
    id: int
    colors: list[Colors] = field(default_factory=list)

    def mincubes(self):
        return Colors(
            red=max(c.red for c in self.colors),
            green=max(c.green for c in self.colors),
            blue=max(c.blue for c in self.colors)
        )


def parse_color(rgb_text):
    res = {}
    for color in rgb_text.split(','):
        value, color = color.split()
        res[color] = int(value)
    return res.get('red', 0), res.get('green', 0), res.get('blue', 0)


def parse_colors(colors):
    for rgb_text in colors.split(';'):
        r, g, b = parse_color(rgb_text)
        yield Colors(
            red=r,
            green=g,
            blue=b
        )


def parse(line):
    game, colors = line.split(':')
    gameid = int(game[len('Game '):])
    return Game(id=gameid, colors=list(parse_colors(colors)))


def read_input(fname):
    for line in open(fname):
        if not line.strip():
            continue
        yield parse(line)
    

fname = 'input.txt'
print(sum(abs(game.mincubes()) for game in read_input(fname)))  # 78669
